### PR TITLE
[task/APPC-2802] Promo code message with App Name

### PR DESCRIPTION
**Pull request by franciscoaleixo**
**What does this PR do?**

   Adds a new message to when a user enters a promo code with bonus only for a specific app. Also improves and refactors parts of the code.

**Database changed?**

   No

**How should this be manually tested?**

  Test the promo code feature and insert a code with and without a specific app bonus to see the different messages.

**What are the relevant tickets?**

  [APPC-2802](https://aptoide.atlassian.net/browse/APPC-2802)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

### CODE REVIEW

File: app/src/main/java/com/asfoundation/wallet/my_wallets/change_wallet/ChangeActiveWalletDialogFragment.kt [hunk 1]
```diff
@@ -52,12 +53,32 @@ class ChangeActiveWalletDialogFragment : DaggerBottomSheetDialogFragment(),
   override fun getTheme(): Int = R.style.AppBottomSheetDialogTheme
 
   override fun onStateChanged(state: ChangeActiveWalletState) {
-    setWalletDetails(state.walletBalance)
+    when (state.changeWalletAsync) {
+      Async.Uninitialized -> setWalletDetails(state.walletBalance)
+      is Async.Loading -> setLoading()
+      is Async.Fail -> Unit
+      is Async.Success -> Unit
```
Comment (danfbfern) [inline app/src/main/java/com/asfoundation/wallet/my_wallets/change_wallet/ChangeActiveWalletDialogFragment.kt:60]: shouldn't Success be used to hide the bottomSheet after the change?

File: app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetFragment.kt [hunk 2]
```diff
@@ -186,8 +179,15 @@ class PromoCodeBottomSheetFragment : DaggerBottomSheetDialogFragment(),
...
-    views.promoCodeBottomSheetSuccessSubtitle.text =
-        this.getString(R.string.promo_code_success_body, bonus.toString())
+    if (promoCode.appName != null) {
+      views.promoCodeBottomSheetSuccessSubtitle.text =
+          this.getString(R.string.promo_code_success_body_specific_app, promoCode.bonus.toString(),
+              promoCode.appName)
+    } else {
+      views.promoCodeBottomSheetSuccessSubtitle.text =
+          this.getString(R.string.promo_code_success_body, promoCode.bonus.toString())
+    }
```
Comment (danfbfern) [inline app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetFragment.kt:189]: I don't know if you agree with this, but I believe putting the setText before the if seems cleaner and more in line with the lines above this.
image

File: app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetViewModel.kt [hunk 3]
```diff
@@ -36,53 +34,39 @@ class PromoCodeBottomSheetViewModel(
   }
 
   private fun getCurrentPromoCode() {
-    observeCurrentPromoCodeUseCase()
-        .asAsyncToState() {
-          copy(promoCodeAsync = it)
-        }
-        .scopedSubscribe() { e ->
+    observeCurrentPromoCodeUseCase().asAsyncToState() {
+      copy(promoCodeAsync = it)
```
Comment (danfbfern) [inline app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetViewModel.kt:38]: I don't find this code style cleaner to read.
I think `asAsyncToState` inline with the `repeatableScopedSubscribe` easier to understand quickly what's going on.

File: app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetViewModel.kt [hunk 4]
```diff
@@ -36,53 +34,39 @@ class PromoCodeBottomSheetViewModel(
...
   }
 
   fun submitClick(promoCodeString: String) {
-    setPromoCodeUseCase(promoCodeString)
-        .asAsyncToState() {
-          copy(submitClickAsync = it)
-        }
-        .scopedSubscribe() { e ->
+    setPromoCodeUseCase(promoCodeString).asAsyncToState() {
+      copy(submitClickAsync = it)
```
Comment (danfbfern) [inline app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetViewModel.kt:47]: same as above

File: app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetViewModel.kt [hunk 5]
```diff
@@ -36,53 +34,39 @@ class PromoCodeBottomSheetViewModel(
...
-  }
+  fun replaceClick() = setState { copy(shouldShowDefault = true) }
 
   fun deleteClick() {
-    deletePromoCodeUseCase()
-        .asAsyncToState() {
-          copy(promoCodeAsync = Async.Uninitialized)
-        }
+    deletePromoCodeUseCase().asAsyncToState() {
+      copy(promoCodeAsync = Async.Uninitialized)
```
Comment (danfbfern) [inline app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetViewModel.kt:58]: same as above

**Review by danfbfern — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/asfoundation/wallet/my_wallets/change_wallet/ChangeActiveWalletDialogFragment.kt (hunk 1 above)
Comment (franciscoaleixo) [reply app/src/main/java/com/asfoundation/wallet/my_wallets/change_wallet/ChangeActiveWalletDialogFragment.kt:60]: This is already done on ViewModel through a side effect

File: app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetFragment.kt (hunk 2 above)
Comment (franciscoaleixo) [reply app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetFragment.kt:189]: Normally, I'd agree for simple if / else statements, but this one is so big that I don't think there's value in it. I guess we could discuss this in code style meeting.

File: app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetViewModel.kt (hunk 3 above)
Comment (franciscoaleixo) [reply app/src/main/java/com/asfoundation/wallet/promo_code/bottom_sheet/PromoCodeBottomSheetViewModel.kt:38]: I agree, it's just code style changes (I had an option enabled that retroactively fixed line breaks -- normally this is disabled).

**Pull request merged**